From eed7c1c7bc4f4d2e343af770504e2c30977de26d Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Fri, 25 Jan 2019 12:34:18 +0530
Subject: [PATCH 0141/1852] ASoC: xlnx: enable multi sample rate support for
 SPDIF sound card

commit 2813b7d6e0e21c2843a6de5737eeb8ce9d43f1ff from
https://github.com/Xilinx/linux-xlnx.git

Patch enables SPDIF sound card to support multiple sampling rates
for playback and capture.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_pl_snd_card.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/sound/soc/xilinx/xlnx_pl_snd_card.c b/sound/soc/xilinx/xlnx_pl_snd_card.c
index 9950ff9b38b5..47177a76bd6b 100644
--- a/sound/soc/xilinx/xlnx_pl_snd_card.c
+++ b/sound/soc/xilinx/xlnx_pl_snd_card.c
@@ -51,6 +51,18 @@ static const char *dev_compat[][XLNX_MAX_IFACE] = {
 	},
 };
 
+static int xlnx_spdif_card_hw_params(struct snd_pcm_substream *substream,
+				     struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct pl_card_data *prv = snd_soc_card_get_drvdata(rtd->card);
+	u32 sample_rate = params_rate(params);
+
+	/* mclk must be >=1024 * sampleing rate */
+	prv->mclk_val = 1024 * sample_rate;
+	prv->mclk_ratio = 1024;
+	return clk_set_rate(prv->mclk, prv->mclk_val);
+}
 
 static int xlnx_sdi_card_hw_params(struct snd_pcm_substream *substream,
 				   struct snd_pcm_hw_params *params)
@@ -178,6 +190,10 @@ static const struct snd_soc_ops xlnx_hdmi_card_ops = {
 	.hw_params = xlnx_hdmi_card_hw_params,
 };
 
+static const struct snd_soc_ops xlnx_spdif_card_ops = {
+	.hw_params = xlnx_spdif_card_hw_params,
+};
+
 SND_SOC_DAILINK_DEFS(xlnx_i2s,
 		     DAILINK_COMP_ARRAY(COMP_CPU("xilinx-i2s")),
 		     DAILINK_COMP_ARRAY(COMP_DUMMY()),
@@ -246,10 +262,12 @@ static struct snd_soc_dai_link xlnx_snd_dai[][XLNX_MAX_PATHS] = {
 		{
 			.name = "xilinx-spdif_playback",
 			SND_SOC_DAILINK_REG(xlnx_spdif),
+			.ops = &xlnx_spdif_card_ops,
 		},
 		{
 			.name = "xilinx-spdif_capture",
 			SND_SOC_DAILINK_REG(xlnx_spdif),
+			.ops = &xlnx_spdif_card_ops,
 		},
 	},
 
-- 
2.31.1

